### ISSUE TITLE
CF: RecommendationAccuracySVDPPTest Fails

### ISSUE THREAD

**Issue by wenhaoh2** [mlpack/mlpack#1501]
`SVDPlusPlus` is a recently implemented model for collaborative filtering. However, the test of recommendation accuracy for `SVDPlusPlus` keeps failing. This issue is for someone who is interested to investigate the reason why this test fails and correct this test.

As of now there is a function `RecommendationAccuracy()` in [cftest.cpp](https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/cf_test.cpp) which tests the recommendation accuracy for a given factorizer. What this test basically does is that, it removes 50 5-star ratings from training dataset, and makes sure that the items corresponding to the removed ratings exist in the top-150 recommendations obtained by `cf.GetRecommendations()`. Even though `SVDPlusPlus` passed the `CFPredictSVDPPTest` which tests the RMSE, it fails `SVDPlusPlus`.

Some ideas to investigate the underlying cause, from my perspectives, are:

1. The default hyperparameters or settings for `SVDPlusPlus` model are not suitable. For example, we might need to tune the step size, regularization factor, rank, etc. We can also try other methods of data normalization, neighbor search and weight interpolation (e.g. Use `OverallMeanNormalization` instead of `NoNormalization`).
2. The test algorithm needs to be improved. For example, the testing might be more reasonable if we ignore all other 5-rating items in the recommendation list before computing the rank of the items corresponding to the removed ratings.
3. The dataset we use for testing CF is too sparse for `SVDPlusPlus` and it is hard to generalize. If we find some relevant evidence that the dataset is too  sparse, we can try to find another dataset more suitable for the tests.

**Comment by utkarshch9** [comment]
i want to contribute in this issue so could you please tell me what more should i do ?

**Comment by rcurtin** [comment]
Hi @utkarshch9, I think that Wenhao has laid out the details of what to do pretty well.  If you're unfamiliar with the code, I'd suggest spending a good amount of time reviewing the code and any relevant papers until you're able to understand what needs to be improved. :+1:

**Comment by darthgera123** [comment]
is this issue still open? if yes i would like to resolve this one

**Comment by rcurtin** [comment]
The issue is not closed, so, yes, it is still open.  If you make a working fix, you can open a PR and we will review it.

**Comment by Sirish07** [comment]
Hey @rcurtin @wenhaoh2 .

I tried resolving the issue as per your suggestion (#3106). However, the test passes on some platforms but fails on the other. Is there any reason behind this behaviour? Thank you.

**Comment by rcurtin** [comment]
@Sirish07 sorry for the slow response.  I'll take a look at the PR when I have a chance, but the most likely reason the test passes on some platforms but fails on others is that the test is a random test (e.g. not deterministic), so when it runs on different platforms the random seed is different, which ends up giving a different result.  Hopefully that helps! :+1:
